core/Introspect.carp

```diff
@@ -4,75 +4,94 @@
 (defmodule Introspect
   (doc module?
     "Is this binding a module?")
-  (defmacro module? [binding]
-    (not (list? (eval (list 's-expr binding)))))
-  (doc function? 
+  (defndynamic module? [binding]
+    (let [s (s-expr binding)]
+      (if (empty? s)
+          false
+          (Dynamic.= (Symbol.from "defmodule") (car s)))))
+
+  (doc function?
     "Is this binding a function?")
-  (defmacro function? [binding]
-    (if (empty? (eval (list 's-expr binding)))
-        false
-      (Dynamic.= (Symbol.from "defn") (car (eval (list 's-expr binding))))))
+  (defndynamic function? [binding]
+    (let [s (s-expr binding)]
+      (if (empty? s)
+          false
+          (Dynamic.= (Symbol.from "defn") (car s)))))
+
   (doc variable?
     "Is this binding a variable?")
-  (defmacro variable? [binding]
-    (if (empty? (eval (list 's-expr binding)))
-        false
-      (Dynamic.= (Symbol.from "def") (car (eval (list 's-expr binding))))))
+  (defndynamic variable? [binding]
+    (let [s (s-expr binding)]
+      (if (empty? s)
+          false
+          (Dynamic.= (Symbol.from "def") (car s)))))
+
   (doc type?
     "Is this binding a type?")
-  (defmacro type? [binding]
-    (if (empty? (eval (list 's-expr binding)))
-        false
-        (Dynamic.= (Symbol.from "deftype") (car (eval (list 's-expr binding))))))
-  (doc struct? 
+  (defndynamic type? [binding]
+    (let [s (s-expr binding)]
+      (if (empty? s)
+          false
+          (Dynamic.= (Symbol.from "deftype") (car s)))))
+
+  (doc struct?
     "Is this binding a struct?")
-  (defmacro struct? [binding]
-    (if (empty? (eval (list 's-expr binding)))
-        false
-        (array? (caddr (eval (list 's-expr binding))))))
+  (defndynamic struct? [binding]
+    (let [s (s-expr binding)]
+      (if (empty? s)
+          false
+          (array? (caddr s)))))
+
   (doc sumtype?
     "Is this binding a sumtype?")
-  (defmacro sumtype? [binding]
-    (if (empty? (eval (list 's-expr binding)))
-        false
-        (list? (caddr (eval (list 's-expr binding))))))
-  (doc arity 
+  (defndynamic sumtype? [binding]
+    (let [s (s-expr binding)]
+      (if (empty? s)
+          false
+          (list? (caddr s)))))
```

Comment:
Should we add a length check here (and above in struct)?

Reply:
Will do! Might be nice to have something generic for this since it happens quite often in my experience. I did implement this elsewhere which more or less does the trick:

```
(defndynamic deconstruct [n else f]                                                                                                                                                  
    (fn [xs]                                                                                                                                                                           
      (if (< (length xs) n)                                                                                                                                                            
          else                                                                                                                                                                         
          (f xs))))                                                                                                                                                                    
                                                                                                                                                                                       
  ;; "safe" versions of cdr, cadr, etc.                                                                                                                                                
  (hidden cdr?)                                                                                                                                                                        
  (private cdr?)                                                                                                                                                                       
  (defndynamic cdr? [xs]                                                                                                                                                               
    (let [f (Folders.Consumers.deconstruct 1 '() cdr)]                                                                                                                                 
      (f xs)))                                                                                                                                                                         
  (hidden cddr?)                                                                                                                                                                       
  (private cddr?)                                                                                                                                                                      
  (defndynamic cddr? [xs]                                                                                                                                                              
    (let [f (Folders.Consumers.deconstruct 2 '() cddr)]                                                                                                                                
      (f xs)))  
```

Reply:
Sounds good, but here I think `(or (empty? s) (< (length s) 3))` should be enough, right?